src/interpreter.rs

```diff
@@ -120,3 +129,1207 @@ bitflags::bitflags! {
+    // of the elements.
+    if len_r + len_s + 7 != sig.len() {
+        return false;
+    };
+
+    // Check whether the R element is an integer.
+    if sig[2] != 0x02 {
+        return false;
+    };
+
+    // Zero-length integers are not allowed for R.
+    if len_r == 0 {
+        return false;
+    };
+
+    // Negative numbers are not allowed for R.
+    if sig[4] & 0x80 != 0 {
+        return false;
+    };
+
+    // Null bytes at the start of R are not allowed, unless R would
+    // otherwise be interpreted as a negative number.
+    if len_r > 1 && sig[4] == 0x00 && sig[5] & 0x80 == 0 {
+        return false;
+    };
+
+    // Check whether the S element is an integer.
+    if sig[len_r + 4] != 0x02 {
+        return false;
+    };
+
+    // Zero-length integers are not allowed for S.
+    if len_s == 0 {
+        return false;
+    };
+
+    // Negative numbers are not allowed for S.
+    if sig[len_r + 6] & 0x80 != 0 {
+        return false;
+    };
+
+    // Null bytes at the start of S are not allowed, unless S would otherwise be
+    // interpreted as a negative number.
+    if len_s > 1 && sig[len_r + 6] == 0x00 && sig[len_r + 7] & 0x80 == 0 {
+        return false;
+    };
+
+    true
+}
+
+fn is_low_der_signature(vch_sig: &ValType) -> Result<bool, ScriptError> {
+    if !is_valid_signature_encoding(vch_sig) {
+        return set_error(ScriptError::SigDER);
+    };
+    // https://bitcoin.stackexchange.com/a/12556:
+    //     Also note that inside transaction signatures, an extra hashtype byte
+    //     follows the actual signature data.
+    let vch_sig_copy = vch_sig.clone();
+    // If the S value is above the order of the curve divided by two, its
+    // complement modulo the order could have been used instead, which is
+    // one byte shorter when encoded correctly.
+    // FIXME: This can return `false` without setting an error, which is not the expectation of the
+    //        caller.
+    Ok(PubKey::check_low_s(&vch_sig_copy))
+}
+
+fn is_defined_hashtype_signature(vch_sig: &ValType) -> bool {
+    if vch_sig.is_empty() {
+        return false;
+    };
+
+    HashType::from_bits(i32::from(vch_sig[vch_sig.len() - 1]), true).is_ok()
+}
+
+fn check_signature_encoding(
+    vch_sig: &Vec<u8>,
+    flags: VerificationFlags,
+) -> Result<bool, ScriptError> {
+    // Empty signature. Not strictly DER encoded, but allowed to provide a
+    // compact way to provide an invalid signature for use with CHECK(MULTI)SIG
+    if vch_sig.is_empty() {
+        return Ok(true);
+    };
+    if !is_valid_signature_encoding(vch_sig) {
+        return set_error(ScriptError::SigDER);
+    } else if flags.contains(VerificationFlags::LowS) && !is_low_der_signature(vch_sig)? {
+        // TODO: The C++ impl claims that `serror` is set here, but it isn’t, so this eventually
+        //       ends up as `UnknownError`, when it should probably be `SigHighS`.
+        return Ok(false);
```

Comment:
You are correct. If `(flags & SCRIPT_VERIFY_LOW_S) != 0 && !IsValidSignatureEncoding(vchSig)`, then `IsLowDERSignature` would have set `serror` to `SCRIPT_ERR_SIG_DER`, but that case is already covered by lines 401-402 and so can't occur. Otherwise `serror` will be unset, when it should be `SCRIPT_ERR_SIG_HIGH_S`.

Reply:
I believe it is safe to fix this bug without causing any backward compatibility problem or consensus divergence, since Bitcoin Core did so in 2016 by https://github.com/bitcoin/bitcoin/pull/7500.